core/filelocation.cpp

```diff
@@ -0,0 +1,235 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "core/filelocation.h"
+#include "core/pref.h"
+#include <QObject>
+#include <QFileInfo>
+#include <QDir>
+#include <utility>
+
+FileLocation currentFile;
+
+FileLocation::FileLocation() : type(NONE)
+{
+}
+
+bool isRemoteRepository(const QString &name)
+{
+	return name.startsWith("http://") || name.startsWith("https://") ||
+	       name.startsWith("ssh://") || name.startsWith("git://");
+}
+
+// For git repositories, remove "file://" and extract username, if any.
+// For local files or git repositories, do proper unicode encoding.
+void FileLocation::fixName()
+{
+	switch (type) {
+	default:
+	case NONE:
+		return;
+	case LOCAL_FILE:
+		name = QFile::encodeName(QDir::toNativeSeparators(name));
+		return;
+	case CLOUD_GIT:
+	case CLOUD_GIT_OFFLINE:
+	case GIT:
+		if (name.startsWith("file://", Qt::CaseInsensitive)) {
+			// Treat "file://..." URLs as local repositories.
+			name = name.mid(7);
+			name = QFile::encodeName(QDir::toNativeSeparators(name));
+		} else if(!isRemoteRepository(name)) {
+			name = QFile::encodeName(QDir::toNativeSeparators(name));
+		} else if(name.startsWith("https://", Qt::CaseInsensitive)) {
+			int at = name.indexOf('@');
+			int slash = name.indexOf('/', 8);
+			if (at >= 0 && slash >= 0 && slash > at) {
+				user = name.mid(8, at - 8);
+				name.remove(8, at - 7);
+			}
+		}
+	}
+}
+
+// Check if filename is of the form "repository[branch]" or "repository[branch].ssrf".
+static bool isGitFilename(const QString &name)
+{
+	QString trimmed = name.trimmed();
+	if (trimmed.endsWith(".ssrf", Qt::CaseInsensitive))
+		trimmed.truncate(trimmed.count() - 5);
+	return trimmed.right(1) == "]" &&
+	       trimmed.lastIndexOf('[') > 0;
+}
+
+// Returns (repository, branch) pair.
+static std::pair<QString, QString> parseGitFilename(const QString &fn)
+{
+	if (!isGitFilename(fn))							// Oops. Filename doesn't follow the rules. Leave branch empty.
```

Comment:
please move the comment closer to the code - this "align on right" is harder to read